src/PostgREST/ApiRequest.hs

```diff
@@ -278,21 +279,21 @@ payloadAttributes raw json =
   -- Test that Array contains only Objects having the same keys
   case json of
     JSON.Array arr ->
-      let objs :: V.Vector JSON.Object
-          objs = foldr -- filter non-objects, map to raw objects
-                   (\val result -> case val of
-                      JSON.Object o -> V.cons o result
-                      _ -> result)
-                   V.empty arr
-          keysPerObj = V.map (S.fromList . M.keys) objs
-          canonicalKeys = fromMaybe S.empty $ keysPerObj V.!? 0
-          areKeysUniform = all (==canonicalKeys) keysPerObj
-          arrLength = V.length arr in
-      if (V.length objs == arrLength) && areKeysUniform
-        then Just $ PayloadJSON raw (PJArray arrLength) canonicalKeys
-        else Nothing
+      case arr V.!? 0 of
+        Just (JSON.Object o) ->
+          let canonicalKeys = M.keys o
+              areKeysUniform = all (\case
+                JSON.Object x -> M.keys x == canonicalKeys
```

Comment:
Your `payloadAttributes` code seems more sensible than the original, less building intermediate variables.

One question, might the comparison `M.keys x == canonicalKeys` yield false negatives sometimes because of the order of the keys in the list? Internally `M.keys` calls `toList` which has a comment in the source code saying "The order of its elements is unspecified."

Reply:
That's true. Though I can't come up with a failing test, I'll just correct that.